### PR TITLE
[WIP] MXNet Extensions enhancements

### PR DESCRIPTION
**Pull request by samskalicky**
## Description ##
Enhancements to MXNet Extensions (subgraph property, custom ops, custom graph passes). Addresses a few points from #17236. More description on the way!

### Features ###
- Enhancement to supportedOps to allow for graph coloring (specifying which subgraph an op should be partitioned into)
- Enhancement for partitioners to use Selector class instead of supportedOps
- Added a new example for custom subgraph op using selector class
- Clean up lib_api.h and library loading in c_api.cc
- Added option to quiet library loading printing ops, partitioners, passes found in library
- Add support for custom graph passes, and docs/README
- Add a new example for custom pass libraries with 2 example passes 
- Added support for allocating new args/aux within the pass, or replacing existing args/aux
- Add new custom pass lib to Makefile/CMakeLists.txt
- Add building sparse custom op libraries to Makefile/CMakeLists.txt
- Add compiling custom libs with lowest supported C++11, building with C++17 for testing with MXNet
- 

#### SupportedOps Enhancements for graph coloring

In the custom partitioner API, custom library writers could implement the `SupportedOps` API to specify which ops to include in a subgraph by setting `True/False` for each node ID in the graph. This PR adds support for writers to identify which specific subgraph to include a node into by specifying a subgraph ID integer for each node, or -1 to indicate that a node can go into any subgraph. 

#### Selector class for subgraph creation

In the custom partitioner API, custom library writers could implement the `SupportedOps` API to specify which ops to include in a subgraph. But some custom partitioners may want more control. This PR, adds support for the internal MXNet subgraph property `SubgraphSelector` class to be implemented in a custom library. Custom library writers can choose to implement `SupportedOps` API or `CustomOpSelector` class for their partitioner. 

#### Custom Graph Passes

This PR adds the ability to register a custom graph pass in a library. Working backwards from the custom library writer, the `Pass` API is implemented with the following interface:

```
MXReturnValue myPass(const std::string& in_graph, const std::string** out_graph,
                     const std::unordered_map<std::string, std::string>& options,
                     const std::unordered_map<std::string, MXTensor>& args,
                     const std::unordered_map<std::string, MXTensor>& aux,
                     const PassResource& res)
```

The model graph is passed as a JSON string, and users provide a new JSON string for the modified graph. Options specified by the MXNet user at the Python level are passed in the `options` map. If the MXNet user provided the args/aux at the Python level they are provided to the pass in the `args/aux` arguments. The PassResource class `res` exposes two functions: `alloc_arg` and `alloc_aux` that allow the custom library writer to allocate NDArrays for new or replacement args/aux within the custom pass. Custom passes are registered in the library with this syntax:

```
REGISTER_PASS(myPassName)
.setBody(myPassFunc);
```

As custom passes are found in the library during loading, a lambda function is registered:

```
auto pass_lambda = [=] (nnvm::Graph&& g) {
    ...
    CHECK(callGraphPass(...));
    ...
    return out_graph;
};

nnvm::PassFunctionReg& pass = dmlc::Registry<nnvm::PassFunctionReg>::Get()->__REGISTER__(myPassName);
pass.set_body(pass_lambda);
pass.set_change_graph(true);
```

From the MXNet front-end, users call custom passes using the same `optimize_for` API. The backend name argument to the `optimize_for` API is attempted to be looked up in registered subgraph backends first, and if missing is attempted to be looked up in registered graph passes:

```
if (mxnet::op::SubgraphBackendRegistry::Get()->backend_map_.count(backend_name) > 0) {
    // use subgraph backend
    ...
} else if (dmlc::Registry<nnvm::PassFunctionReg>::Find(backend_name) != nullptr) {
    // use graph pass
    ...
    g = ApplyPass(std::move(g), backend_name);
    ...
}
```

#### Compilation
The lib_api.h was designed to be as generic as possible to allow users to compile their custom library with any version of C++ (C++11 or higher) and GLIBC to fit the needs of their application. This does not have to correspond to the version of C++ or GLIBC used to compile MXNet. To test this, we will compile with C++11 to check for compiler errors, but will also compile and test with C++17 to match was MXNet is using. 

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by samskalicky** [comment]
@mxnet-bot run ci [sanity]

**Comment by samskalicky** [comment]
Jenkins CI successfully triggered : [sanity]

**Comment by samskalicky** [comment]
@mxnet-bot run ci [sanity]

**Comment by samskalicky** [comment]
@mxnet-bot run ci [sanity]

**Comment by samskalicky** [comment]
@mxnet-bot run ci [sanity]

File: CMakeLists.txt [hunk 1]
```diff
@@ -750,17 +750,33 @@ endif()
 
 # extension libraries (custom operators, custom subgraphs) are built by default
 add_library(customop_lib SHARED ${CMAKE_CURRENT_SOURCE_DIR}/example/extensions/lib_custom_op/gemm_lib.cc)
+add_library(transposecsr_lib SHARED ${CMAKE_CURRENT_SOURCE_DIR}/example/extensions/lib_custom_op/transposecsr_lib.cc)
+add_library(transposerowsp_lib SHARED ${CMAKE_CURRENT_SOURCE_DIR}/example/extensions/lib_custom_op/transposerowsp_lib.cc)
```
Comment (marcoabreu) [inline CMakeLists.txt:754]: Is there some way to automatically discover all the .cc files in the lib_custom_op folder instead of hardcoding all the paths?

File: example/extensions/lib_custom_op/gemm_lib.cc [hunk 2]
```diff
@@ -136,13 +137,13 @@ MXReturnValue inferType(std::map<std::string, std::string> attrs,
...
-  outtypes[0] = intypes[0];
+  outtypes->at(0) = intypes[0];
   return MX_SUCCESS;
 }
 
-MXReturnValue inferShape(std::map<std::string, std::string> attrs,
-                         std::vector<std::vector<unsigned int>> &inshapes,
-                         std::vector<std::vector<unsigned int>> &outshapes) {
+MXReturnValue inferShape(const std::unordered_map<std::string, std::string>& attrs,
+                         const std::vector<std::vector<unsigned int>>& inshapes,
```
Comment (ptrendx) [inline example/extensions/lib_custom_op/gemm_lib.cc:145]: This should not be a const ref - the operator should be able to infer shapes in the other direction (output -> input) as well.

File: example/extensions/lib_custom_op/relu_lib.cu [hunk 3]
```diff
@@ -29,93 +29,93 @@
...
+  return MX_SUCCESS;
 }
 
-MXReturnValue inferType(std::map<std::string, std::string> attrs,
-                        std::vector<int> &intypes,
-                        std::vector<int> &outtypes) {
-    outtypes[0] = intypes[0];
-    return MX_SUCCESS;
+MXReturnValue inferType(const std::unordered_map<std::string, std::string>& attrs,
+                        const std::vector<int>& intypes,
```
Comment (ptrendx) [inline example/extensions/lib_custom_op/relu_lib.cu:108]: Same as in infershape.

File: include/mxnet/lib_api.h [hunk 4]
```diff
@@ -507,16 +553,13 @@ class OpResource {
...
 /*! \brief Macro to help passing serialized subgraph through attribute dict */
 #define MX_STR_SUBGRAPH_SYM_JSON "subgraph_sym_json"
-#define MX_STR_DTYPE "__dtype__"
-#define MX_STR_SHAPE "__shape__"
+#define MX_STR_DTYPE "__ext_dtype__"
+#define MX_STR_SHAPE "__ext_shape__"
 
 /* \brief get shape value from list of shapes string
- * format: [[1]] or [[1],[2]]
+ * format: [[1]] or [[1],[2,3]], returns "[1]" or "[2,3]"
```
Comment (ptrendx) [inline include/mxnet/lib_api.h:562]: This docstring is confusing. It should have the index mentioned there (maybe something like:
`getShapeAt("[[1]]", 0) returns "[1]"` and `getShapeAt("[[1],[2,3]]", 1) returns "[2,3]"`).

File: include/mxnet/lib_api.h [hunk 5]
```diff
@@ -529,7 +572,7 @@ std::string getShapeAt(const std::string& shape, unsigned index) {
 }
 
 /* \brief get dtype value from list of dtypes string
- * format: [1] or [1,2]
+ * format: [1] or [1,2], returns "1" or "2"
```
Comment (ptrendx) [inline include/mxnet/lib_api.h:575]: Same as above.

File: Makefile [hunk 6]
```diff
@@ -686,21 +686,36 @@ pylint:
...
 extension_libs: $(EXT_LIBS)
 
 build/libcustomop_lib.so:
 	@mkdir -p $(@D)
+	$(CXX) -shared -fPIC -std=c++11 example/extensions/lib_custom_op/gemm_lib.cc -o /dev/null -I include/mxnet
 	$(CXX) -shared -fPIC -std=c++17 example/extensions/lib_custom_op/gemm_lib.cc -o $@ -I include/mxnet
+build/libtransposecsr_lib.so:
+	@mkdir -p $(@D)
+	$(CXX) -shared -fPIC -std=c++11 example/extensions/lib_custom_op/transposecsr_lib.cc -o /dev/null -I include/mxnet
+	$(CXX) -shared -fPIC -std=c++17 example/extensions/lib_custom_op/transposecsr_lib.cc -o $@ -I include/mxnet
```
[the remark this answers is not in the document]
Comment (rondogency) [reply Makefile:702]: plus since this will only go into 1.7 where c++17 is not included, we don't need this test at all

**Comment by rondogency** [comment]
also please lib_api update version to 10

File: example/extensions/lib_custom_op/gemm_lib.cc [hunk 7]
```diff
@@ -87,20 +87,20 @@ MXReturnValue forward(std::map<std::string, std::string> attrs,
  ***** gradient outputs
  * outputs[0] = dA; outputs[1] = dB
  */
-MXReturnValue backward(std::map<std::string, std::string> attrs,
-                       std::vector<MXTensor> inputs,
-                       std::vector<MXTensor> outputs,
-                       OpResource res) {
+MXReturnValue backward(const std::unordered_map<std::string, std::string>& attrs,
+                       std::vector<MXTensor>* inputs,
```
[the remark this answers is not in the document]
Comment (samskalicky) [reply example/extensions/lib_custom_op/gemm_lib.cc:91]: done

**Comment by samskalicky** [comment]
> also please lib_api update version to 10

changed to version 7

File: CMakeLists.txt [hunk 8]
```diff
@@ -726,18 +726,39 @@ endif()
...
+target_include_directories(transposecsr_lib PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/include/mxnet)
+target_include_directories(transposerowsp_lib PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/include/mxnet)
 target_include_directories(subgraph_lib PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/include/mxnet)
+target_include_directories(pass_lib PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/include/mxnet)
 if(USE_CUDA)
   add_library(customop_gpu_lib SHARED ${CMAKE_CURRENT_SOURCE_DIR}/example/extensions/lib_custom_op/relu_lib.cu)
   target_include_directories(customop_gpu_lib PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/include/mxnet)
 endif()
-if(MSVC)
+if(UNIX)
```
[the remark this answers is not in the document]
Comment (samskalicky) [reply CMakeLists.txt:742]: done

File: include/mxnet/lib_api.h (hunk 5 above)
Comment (samskalicky) [reply include/mxnet/lib_api.h]: done

File: include/mxnet/lib_api.h (hunk 4 above)
Comment (samskalicky) [reply include/mxnet/lib_api.h]: done

File: example/extensions/lib_custom_op/relu_lib.cu (hunk 3 above)
Comment (samskalicky) [reply example/extensions/lib_custom_op/relu_lib.cu]: done

File: example/extensions/lib_custom_op/gemm_lib.cc (hunk 2 above)
Comment (samskalicky) [reply example/extensions/lib_custom_op/gemm_lib.cc]: done

File: CMakeLists.txt (hunk 8 above)
[the remark this answers is not in the document]
Comment (rondogency) [reply CMakeLists.txt:742]: I mean you don't need to add -shared even for customop_gpu_lib, just change the lines inside MSVC block

File: example/extensions/lib_custom_op/README.md [hunk 9]
```diff
@@ -22,15 +22,13 @@ C++ Custom Operator Example and Tutorial
...
+Custom operators (CustomOp) enable users to write new operators without compiling against all of MXNet header files and dependencies. When a library containing custom operators is loaded dynamically, the operators found in the library will be registered in MXNet so that users can call those operators natively just like other built-in operators.
 
 ## Getting Started
 
 ### Have MXNet Ready
 
-Custom Operator support was merged (#15921, #17270) and is not available in versions of MXNet prior to v1.7.0.
-To access the feature now, please install MXNet by compiling from source using master or using the previously mentioned commits, downloading one of the nightly builds, or from a release of MXNet 1.7.0+.
-For running the following example, it doesn’t matter if it is a CUDA, MKLDNN or plain MXNet build; the custom operator doesn’t interact with the execution of other native MXNet operators.
+To run the following example, the build type of MXNet doesn’t matter since the custom operator doesn’t interact with the execution of other native MXNet operators.
```
Comment (rondogency) [inline example/extensions/lib_custom_op/README.md:31]: it is better to add prerequisite here or run an example like "This requires GCC > 5 or CUDA > 9 to run the examples"

File: include/mxnet/lib_api.h [hunk 10]
```diff
@@ -22,7 +22,11 @@
  * \file lib_api.h
  * \brief APIs to interact with libraries
  * This API specifies function prototypes to
- * register custom ops for library authors
+ * register custom ops, partitioner, and passes
+ * for library authors
+ * See example/extension/lib_custom_op/README.md
```
Comment (rondogency) [inline include/mxnet/lib_api.h:27]: maybe we can rephrase it to "APIs to write extension library, see ... for registering custom operators, ... for custom partitioners, ... for custom graph passes"

File: include/mxnet/lib_api.h [hunk 11]
```diff
@@ -45,7 +49,7 @@
 #endif
 
 /* Make sure to update the version number everytime you make changes */
-#define MX_LIBRARY_VERSION 6
+#define MX_LIBRARY_VERSION 7
```
Comment (rondogency) [inline include/mxnet/lib_api.h:52]: I still feel it is better to make it 10

also it is better to add to c_api.cc line 339 version checking message something like "please update lib_api.h to match the version supported by MXNet backend"

File: example/extensions/lib_custom_op/README.md [hunk 12]
```diff
@@ -200,6 +208,9 @@ If the number of input and output tensors are fixed, you can use hard-coded numb
 * **inferType**: This function takes three arguments. The 1st argument is the attributes (same as above). The 2nd argument is the a list of input data types corresponding to the input tensors. The 3rd argument is the placeholder for output tensor data types you need to assign.
 For example, if this operator has one input and one output, and data type doesn’t change, then you can do `outtypes[0] = intypes[0]` to populate the data type.
 
+* **inferSType**: This function takes three arguments. The 1st argument is the attributes (same as above). The 2nd argument is the a list of input storage types corresponding to the input tensors. The 3rd argument is the placeholder for output storage types you need to assign.
+For example, if this operator has one input and one output, and data type doesn’t change, then you can do `outtypes[0] = intypes[0]` to populate the data type.
```
Comment (eric-haibin-lin) [inline example/extensions/lib_custom_op/README.md:212]: data type doesn’t change -> data storage type doesn’t change

Comment (eric-haibin-lin) [reply example/extensions/lib_custom_op/README.md:212]: a list of input storage types corresponding to the input tensors -> a list of input storage types corresponding to the input tensors (dense, row_sparse, or CSR). For details, see https://cwiki.apache.org/confluence/display/MXNET/A+Guide+to+Implementing+Sparse+Operators+in+MXNet+Backend 

It would be good to include the link above in case people wonder why/if inferSType is needed.

**Comment by ptrendx** [comment]
Synced offline with @samskalicky and @rondogency - they are ok with doing the last changes proposed by @rondogency in another PR targeting master specifically, so this PR is ready to merge.

File: example/extensions/lib_pass/README.md [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+Then compile it to the `mypass_lib.so` dynamic library using the following command:
+```bash
+g++ -shared -fPIC -std=c++11 mypass_lib.cc -o libmypass_lib.so -I ../../../include/mxnet
+```
+
+Finally, you can write a Python script to load the library and execute your pass on a model:
+
+```python
+import mxnet as mx
+mx.library.load(‘libmypass_lib.so’)
```
Comment (eric-haibin-lin) [inline example/extensions/lib_pass/README.md:67]: ‘libmypass_lib.so’  -> 'libmypass_lib.so'

**Comment by leezu** [comment]
@ptrendx let's not merge commits named "[WIP]" to master. You can edit the name prior to merge

**Comment by ptrendx** [comment]
Oops, you are right, sorry, I missed that.

File: example/extensions/lib_custom_op/README.md (hunk 12 above)
Comment (samskalicky) [reply example/extensions/lib_custom_op/README.md:212]: We need a whole overview for Sparse. Maybe you can help us add another section to the readme about that

File: example/extensions/lib_pass/README.md (hunk 13 above)
Comment (samskalicky) [reply example/extensions/lib_pass/README.md:67]: what am i missing? looks the same to me...

**Comment by samskalicky** [comment]
> @ptrendx let's not merge commits named "[WIP]" to master. You can edit the name prior to merge

We didnt want to rerun the whole CI. did we fix the problem where renaming the PR reruns CI @leezu ?

**Comment by leezu** [comment]
< We didnt want to rerun the whole CI. did we fix the problem where renaming the PR reruns CI @leezu ?

@samskalicky when merging the commit, Github allows to edit the commit message. So the commit message can be different from the PR title. Indeed we can't edit the PR title without retriggering the CI as of now (cc @ChaiBapchya)
